search.py

```diff
@@ -509,38 +509,40 @@ def and_search(states, problem, path):
     # body of and or search
     return or_search(problem.initial, problem, [])
 
+# Pre-defined actions for PeakFindingProblem
+primary_actions = { 'W':(-1,0), 'N':(0,1), 'E':(1,0), 'S':(0,-1) }
+aggregate_actions = { 'W':(-1,0), 'N':(0,1), 'E':(1,0), 'S':(0,-1), 'NW':(-1,1), 'NE':(1,1), 'SE':(1,-1), 'SW':(-1,-1) }
 
 class PeakFindingProblem(Problem):
     """Problem of finding the highest peak in a limited grid"""
 
-    def __init__(self, initial, grid):
+    def __init__(self, initial, grid, defined_actions = primary_actions):
         """The grid is a 2 dimensional array/list whose state is specified by tuple of indices"""
         Problem.__init__(self, initial)
         self.grid = grid
+        self.defined_actions = defined_actions
         self.n = len(grid)
         assert self.n > 0
         self.m = len(grid[0])
         assert self.m > 0
 
     def actions(self, state):
-        """Allows movement in only 4 directions"""
-        # TODO: Add flag to allow diagonal motion
+        """Returns the list of actions which are allowed to be taken from the given state"""
         allowed_actions = []
-        if state[0] > 0:
-            allowed_actions.append('N')
-        if state[0] < self.n - 1:
-            allowed_actions.append('S')
-        if state[1] > 0:
-            allowed_actions.append('W')
-        if state[1] < self.m - 1:
-            allowed_actions.append('E')
+        for action in self.defined_actions:
+            next_state = (state[0] + self.defined_actions[action][0], state[1] + self.defined_actions[action][1]) 
+            if next_state[0] >= 0 and next_state[1] >= 0 and next_state[0] <= self.n - 1 and next_state[1] <= self.m - 1:
+                allowed_actions.append(action)
+
         return allowed_actions
 
     def result(self, state, action):
         """Moves in the direction specified by action"""
         x, y = state
```

Comment:
I think we have defined an `add` function for points/vectors. I forget what it is called, but if it is `add` then `result` should be just `return add(state, action)`